8365053: Refresh hotspot precompiled.hpp with headers based on current frequency

fandreuz opened the pull request:
In this PR I propose to refresh the included headers in hotspot `precompiled.hpp`. The current set of precompiled headers was refreshed in 2018, 7 years ago. I repeated the same operations and measurements after refreshing the set of precompiled headers according to the current usage frequency.

These are the results I observed. Depending on the platform, the improvement is between 10 and 20% in terms of total work (user+sys). The results are in seconds.

```
linux-x64 GCC
master      real 81.39 user 3352.15 sys 287.49
JDK-8365053 real 81.94 user 3030.24 sys 295.82

linux-x64 Clang
master      real 43.44 user 2082.93 sys 130.70
JDK-8365053 real 38.44 user 1723.80 sys 117.68

linux-aarch64 GCC
master      real 1188.08 user 2015.22 sys 175.53
JDK-8365053 real 1019.85 user 1667.45 sys 171.86

linux-aarch64 clang
master      real 981.77 user 1645.05 sys 118.60
JDK-8365053 real 791.96 user 1262.92 sys 101.50
```

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8365053](https://bugs.openjdk.org/browse/JDK-8365053): Refresh hotspot precompiled.hpp with headers based on current frequency (**Enhancement** - P4)


### Reviewers
 * [Quan Anh Mai](https://openjdk.org/census#qamai) (@merykitty - Committer) Review applies to [71950aed](https://git.openjdk.org/jdk/pull/26681/files/71950aedacb3bf523c461b76ca1db2bed5816348)
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)
 * [Magnus Ihse Bursie](https://openjdk.org/census#ihse) (@magicus - **Reviewer**)
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/26681/head:pull/26681` \
`$ git checkout pull/26681`

Update a local copy of the PR: \
`$ git checkout pull/26681` \
`$ git pull https://git.openjdk.org/jdk.git pull/26681/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 26681`

View PR using the GUI difftool: \
`$ git pr show -t 26681`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/26681.diff">https://git.openjdk.org/jdk/pull/26681.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/26681#issuecomment-3165028586)
</details>

fandreuz commented:
/label hotspot build

merykitty commented on src/utils/PrecompiledHeaders/PrecompiledHeaders.java:
```diff
@@ -0,0 +1,134 @@
+        try (Stream<Path> paths = Files.walk(hotspotPath)) {
+            paths.filter(Files::isRegularFile)
+                    .filter(path -> {
+                        String name = path.getFileName().toString();
```
IIUC this means that we count the number of cpp or hpp files that directly include a file, but should we count the number of cpp file that directly or transitively include a file instead?

dholmes-ora commented:
> For example, when the threshold is set to 50 I get this compilation error:

How can we possibly get a compilation error when everything should build with PCH disabled?

kimbarrett commented on src/utils/PrecompiledHeaders/PrecompiledHeaders.java:
```diff
@@ -0,0 +1,134 @@
+
+        // Count inclusion times for each header
+        Map<String, Integer> occurrences = new HashMap<>();
+        try (Stream<Path> paths = Files.walk(hotspotPath)) {
```
I think walking the source tree is the wrong approach to gathering the data
about include counts. I think better is to do a build and then look at the
files <builddir>/hotspot/variant-server/libjvm/objs/*.d. From that one can build
a completely accurate count of the transitive inclusions.

fandreuz replied:
I followed this hint, the latest version of the script checks the `.d` files. 

Two observations:
- The magic number is `2460` (more includes are taken into account, which makes sense)
- There is much less wiggle room now, 2461 includes nothing, 2459 includes too much
- Runtime does not seem to be affected negatively or positively

If anybody is interested, this file contains the inclusion count for each source file: [inclusions_count.txt](https://github.com/user-attachments/files/21682561/inclusions_count.txt)

fandreuz replied:
> Note that precompiled.hpp has an include count of 2456. There's 1 .d file that
depends on everything - BUILD_LIBJVM.d. That probably ought to be excluded
from counting, although I guess it should just add one to every file.

Apparently `BUILD_LIBJVM.d` mentions all files multiple times:
```
% grep "utilities/waitBarrier.hpp" build/clang/hotspot/variant-server/libjvm/objs/BUILD_LIBJVM.d | wc -l
1231
```

It's definitely reasonable to exclude it, I did it in the latest revision of the script.

Now, the numbers seem more reasonable:
```
% ls build/clang/hotspot/variant-server/libjvm/objs/*.d | grep -v BUILD | wc -l
1232
% grep "code/codeCache.hpp" build/clang/hotspot/variant-server/libjvm/objs/*.d | \
            grep -v BUILD | sort | uniq | wc -l
1230
% grep "code/codeCache.hpp" inclusions_count.txt
code/codeCache.hpp=1230
```

Updated [inclusions_count.txt](https://github.com/user-attachments/files/21685761/inclusions_count.txt)

merykitty replied:
I assume this means that other `.d` files might include a header multiple times, too. Is it reasonable to count only the number of files containing a header instead of number of appearances?

fandreuz replied:
> I assume this means that other .d files might include a header multiple times, too.

From the following two numbers, I deduce `code/codeCache.hpp` is (transitively) included in 1230 files, with each inclusion coming from a unique `.d` file. Am I missing something?

```
% grep "code/codeCache.hpp" build/clang/hotspot/variant-server/libjvm/objs/*.d | grep -v BUILD | sort | wc -l
1230

% grep "code/codeCache.hpp" build/clang/hotspot/variant-server/libjvm/objs/*.d | grep -v BUILD | sort | uniq | wc -l
1230
```

I haven't checked every single file, but I'd expect each include to appear at most once in each `.d` file.

merykitty replied:
Thanks for the clarification, I think that it is fine, then, although a comment explaining in the code would be great.

kimbarrett replied:
The number of times a header is (directly or indirectly) included by a given
source file doesn't affect the number of times that header appears in the .d
file associated with that source file. The compiler uniquifies that set
(though links might confuse things, but that's not an issue here).

The .d files consist of a set of targets, and for each target, the set of
dependencies.

For the <sourcefile>.d files, there's one target, <sourcefile>.o, with its
dependencies (once each).

For BUILD_LIBJVM.d, there's a target for each .o file. And for each of those
targets, it's dependencies (once each). Essentially BUILD_LIBJVM.d includes a
concatenation of all the <sourcefile>.d files. And since there are duplicates
among the dependencies of different targets, there are many(!) duplicates in
this file.

kimbarrett commented on src/hotspot/share/precompiled/precompiled.hpp:
```diff
@@ -26,52 +26,38 @@
 // These header files are included in at least 130 C++ files, as of
-// measurements made in November 2018. This list excludes files named
-// *.inline.hpp, since including them decreased build performance.
+// measurements made in August 2025.
```
I don't think there's anything particularly special about the number 130.

Another thing to consider when a header file has a high include count is
whether it's being overincluded. We've had lots of those, and some folks
occasionally try to poke at that problem. Some of the removals here look like
they might be a result of such efforts.

Still another thing to consider is the cost of inclusion. Some files may just
be a lot more expensive to process and benefit more for being precompiled.
File size can be an indicator, but there are others. Unfortunately, I don't
know of a good way to measure this.

merykitty replied:
You need to modify the comment here, too.

magicus replied:
This number is based on my original experimentation. If I tried to lower the bar by lowering the number, the PCH list grew too much and it made for a worse performance. And contrary, if I raised the number, fewer files where included which made the PCH quicker to process but less helpful. That number was the optimum I found. It seems from https://bugs.openjdk.org/browse/JDK-8365053 that this is still around the optimum. However, rather than just mentioning the number in the comment, the rationale could be specified, like `the optimum number of includes`.

kimbarrett commented on src/hotspot/share/precompiled/precompiled.hpp:
```diff
@@ -26,52 +26,38 @@
 #include "utilities/ostream.hpp"
-#include "utilities/ticks.hpp"
-
-#ifdef TARGET_COMPILER_visCPP
```
All of the reported testing was on Linux. These were included specifically because measurements
said their inclusion here was beneficial. And my recollection from previous discussions is that
Visual Studio may be the compiler where precompiled headers are most beneficial.

merykitty commented on src/utils/PrecompiledHeaders/PrecompiledHeaders.java:
```diff
@@ -0,0 +1,136 @@
+                    .flatMap(file -> {
+                        try {
+                            // The first line contains the object name
+                            return Files.lines(file).skip(1);
```
Or maybe `return Files.lines(file).skip(1).distince()`?

fandreuz replied:
`distinct()` has some overhead, do we need it? Do we expect duplicate headers in a file?

fandreuz commented:
> The latest version fails for me. I'm guessing because we exclude shenandoah by default for OracleJDK builds so having those headers in precompiled.hpp messes things up. I think we need to be careful with headers that are part of optional features.

The last commit (be25d3413b432b56e9789eae55920f1862008911) should fix the problem. I made a build with the following configuration:
```
./configure \
        --with-toolchain-type=clang \
        --disable-jvm-feature-shenandoahgc \
        --disable-jvm-feature-zero \
        --disable-jvm-feature-zgc \
        --disable-jvm-feature-g1gc \
        --disable-jvm-feature-parallelgc \
        --disable-jvm-feature-epsilongc
```
and then I ran the script on this build to extract the [inclusions count](https://github.com/user-attachments/files/21692915/inclusions_count.txt). This could be included in the README file attached to this PR.

kimbarrett commented:
> > The latest version fails for me. I'm guessing because we exclude shenandoah by default for OracleJDK builds so having those headers in precompiled.hpp messes things up. I think we need to be careful with headers that are part of optional features.
> 
> The last commit ([be25d34](https://github.com/openjdk/jdk/commit/be25d3413b432b56e9789eae55920f1862008911)) should fix the problem. I made a build with the following configuration:
> 
> ```
> ./configure \
>         --with-toolchain-type=clang \
>         --disable-jvm-feature-shenandoahgc \
>         --disable-jvm-feature-zero \
>         --disable-jvm-feature-zgc \
>         --disable-jvm-feature-g1gc \
>         --disable-jvm-feature-parallelgc \
>         --disable-jvm-feature-epsilongc
> ```
> 
> and then I ran the script on this build to extract the [inclusions count](https://github.com/user-attachments/files/21692915/inclusions_count.txt). This could be included in the README file attached to this PR.

Other optional features include cds, c1, jfr, and jvmci, so files from those
directories also ought to be excluded.

The (current) list of optional features is here:
https://github.com/openjdk/jdk/blame/022e29a77533aacabd56820d00ecffa9646a8362/make/autoconf/jvm-features.m4#L47-L49

    cds compiler1 compiler2 dtrace epsilongc g1gc jfr jni-check \
    jvmci jvmti link-time-opt management minimal opt-size parallelgc \
    serialgc services shenandoahgc vm-structs zero zgc \

Not all of these have their own directories.  Even where they do, the feature
name and directory name aren't always the same. (The compiler1 feature and the
c1 directory being an obvious example.)

I didn't spot any files for other optional features in the latest list, other
than the ones called out in the first sentence of this comment.

magicus commented:
There is a handy shortcut for creating a build with all optional features disabled, `--with-jvm-variants=custom`.

fandreuz commented:
> > But still need to figure out which files to (conditionally) exclude.
> 
> Maybe first make a baseline build with no optional features, and analyze the result with this tool. And then enable each individual feature, re-build and re-analyze, and see how it differs. If a file then makes it to the PCH list, it is useful when enabling that feature, and it is potentially only correct to include with that feature enabled, so it could be added with an include guard.
> 
> That will require a bit more manual work (unless this process too can be automated), but it will probably give the best results.

Hi @magicus, thanks for the hint. I'll give it a shot and post the results here!

fandreuz commented:
| Feature               | Baseline `precompiled.hpp` (Mean ±)  | Updated `precompiled.hpp` (Mean ±)  |
|-----------------------|---------------------|----------------------|
| epsilongc             | 505.78 ± 9.08        | 499.36 ± 3.21         |
| epsilongc cds         | 523.93 ± 9.42        | 510.68 ± 0.99         |
| epsilongc compiler1   | 514.89 ± 2.28        | 518.67 ± 14.22        |
| epsilongc compiler2   | 758.34 ± 5.56        | 736.65 ± 32.95        |
| g1gc                  | 674.96 ± 6.37        | 693.39 ± 23.25        |
| epsilongc services jfr| 707.45 ± 2.08        | 670.55 ± 1.50         |
| epsilongc jni-check   | 522.61 ± 1.82        | 517.46 ± 1.67         |
| epsilongc compiler1 jvmci | 604.92 ± 4.23    | 600.61 ± 8.79         |
| epsilongc compiler2 jvmci | 836.85 ± 7.40    | 792.07 ± 16.18        |
| epsilongc services jvmti | 549.48 ± 6.87     | 553.74 ± 7.67         |
| epsilongc management  | 486.56 ± 5.68        | 480.58 ± 4.59         |
| epsilongc opt-size    | 476.57 ± 1.87        | 476.71 ± 1.23         |
| parallelgc            | 497.28 ± 3.90        | 493.68 ± 2.19         |
| serialgc              | 477.84 ± 0.86        | 476.14 ± 1.32         |
| epsilongc services    | 476.99 ± 2.44        | 478.74 ± 3.47         |
| shenandoahgc          | 676.10 ± 5.35        | 623.43 ± 3.11         |
| zgc                   | 620.95 ± 1.84        | 621.86 ± 4.35         |

magicus commented:
This looks like regressions for c1 and G1..?

fandreuz commented:
I'm starting to think the high variability is a quirk of my environment. I ran the builds in a virtualized machine and I'm probably experiencing erratic behavior under high load. 

I re-ran each combination fewer times (3, should still be enough to get an idea), and I'm observing more reasonable results:

| Feature                  | Baseline `precompiled.hpp`  | Updated `precompiled.hpp`           |
|----------------------------|-----------------|-----------------|
| epsilongc                  | 458.88 +/- 0.35 | 459.69 +/- 0.34 |
| epsilongc cds              | 501.63 +/- 0.35 | 501.58 +/- 0.57 |
| epsilongc compiler1        | 505.66 +/- 0.61 | 505.06 +/- 0.51 |
| epsilongc compiler2        | 725.91 +/- 0.41 | 692.25 +/- 0.40 |
| g1gc                       | 630.73 +/- 0.02 | 624.24 +/- 0.35 |
| epsilongc services jfr     | 625.54 +/- 1.45 | 592.37 +/- 1.00 |
| epsilongc jni-check        | 461.19 +/- 0.14 | 461.79 +/- 0.89 |
| epsilongc compiler1 jvmci  | 541.31 +/- 0.42 | 540.31 +/- 0.70 |
| epsilongc compiler2 jvmci  | 758.13 +/- 0.71 | 724.80 +/- 0.21 |
| epsilongc services jvmti   | 499.64 +/- 0.18 | 499.69 +/- 0.52 |
| epsilongc management       | 450.82 +/- 0.62 | 451.38 +/- 0.59 |
| epsilongc opt-size         | 446.44 +/- 0.56 | 446.16 +/- 0.12 |
| parallelgc                 | 467.23 +/- 0.19 | 466.89 +/- 0.41 |
| serialgc                   | 454.17 +/- 0.40 | 454.22 +/- 0.12 |
| epsilongc services         | 453.82 +/- 0.21 | 453.59 +/- 0.51 |
| shenandoahgc               | 658.17 +/- 0.50 | 647.44 +/- 0.40 |

[This](https://github.com/fandreuz/jdk-stuff/blob/master/script) is the script I used.

fandreuz commented:
> Your updated result seemed more reliable and probable, but they also do not differ that much from the baseline. I still saw a few improvements and no regression, so it might still be worth pursuing, but maybe your suggested change does not bring as much improvement as was first hinted at.

I'd say individual improvements are a bit harder to observe in the kind of measurements I'm doing now, and the likelihood of measuring noise is higher. Also, the minimum inclusion threshold I'm using now is the same for all features, but that's unlikely to be optimal.

Maybe we could revert to the original approach (count inclusion for a typical build)? That proved to improve compile time, and I could hide non-standard headers behind an `#if`.

magicus commented:
If I recall correctly the original approach was even simpler: I counted the number of `#include`s in the C++ source files, not in the dependency files of a particular build. Crude but effective, and left out discussions about differences between platforms and feature configurations. And then I had to do some adjustments for Windows where my approach of excluding inline files did not hold and caused a regression.

I think your approach seems more valid, but maybe we get stuck in details then instead...

xmas92 reviewed (commented):
I have a few questions/comments.

There are now a bunch of unconditional includes for JVM features which can be disabled. With the exception of CDS, in the rest of the codebase we have been very careful to always guard these includes when using them outside their own sub-components. 

I understand that these are generated includes, but I think we can curate the list some manually.

Currently `c1/c1_globals.hpp` `jvmci/jvmci_globals.hpp` and `opto/c2_globals.hpp` are included which are optional features, `compiler/compiler_globals.hpp` (which is also included) should be used instead. Same goes for `gc/serial/serial_globals.hpp` and `gc/shared/gc_globals.hpp`.

In addition to those globals files we have all the JFR and CDS includes which are unguarded.

I am not sure I understand completely what goes into the list of conditional includes here. I understand that there might be duplicates across the different conditional includes, but now there are also includes which are duplicated in the unconditional includes. So is it a list of all includes which are beneficial when building with just that feature, or is it suppose to be a list of extra includes which are beneficial if that feature is enabled. Or are those lists something else?

fandreuz commented:
I think I found a more sensible approach to tackle this problem. Using clang [`-ftime-trace`](https://clang.llvm.org/docs/analyzer/developer-docs/PerformanceInvestigation.html#performance-analysis-using-ftime-trace) we can get reports in [Trace Event format](https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/preview?tab=t.0) about each header. Example of one such file here: [shenandoahOldGC.json](https://github.com/user-attachments/files/21915502/shenandoahOldGC.json).

These files can be processed (e.g. with [ClangBuildAnalyzer](https://github.com/aras-p/ClangBuildAnalyzer/tree/main)) to dig where time was spent during the build. Among the information we can get from `ClangBuildAnalyzer`, here is the interesting one:
```
**** Expensive headers:
597169 ms: /jdk/src/hotspot/share/oops/access.inline.hpp (included 650 times, avg 918 ms), included via:
  80x: oop.inline.hpp iterator.inline.hpp 
  70x: javaClasses.inline.hpp 
  40x: jfrEvents.hpp jfrEventClasses.hpp jfrEvent.hpp jfrNativeEventWriter.hpp jfrEventWriterHost.inline.hpp jfrEventWriterHost.hpp jfrWriterHost.inline.hpp jfrTraceId.inline.hpp javaThread.inline.hpp oopHandle.inline.hpp 
  39x: shenandoahHeap.inline.hpp javaClasses.inline.hpp 
  32x: g1CollectedHeap.inline.hpp g1ConcurrentMark.inline.hpp g1ConcurrentMarkBitMap.inline.hpp markBitMap.inline.hpp oop.inline.hpp iterator.inline.hpp 
  30x: ciUtilities.inline.hpp interfaceSupport.inline.hpp javaThread.inline.hpp oopHandle.inline.hpp 
  ...

425714 ms: /jdk/src/hotspot/share/memory/iterator.inline.hpp (included 646 times, avg 659 ms), included via:
  80x: oop.inline.hpp 
  70x: javaClasses.inline.hpp access.inline.hpp barrierSet.inline.hpp objArrayOop.inline.hpp oop.inline.hpp 
  40x: jfrEvents.hpp jfrEventClasses.hpp jfrEvent.hpp jfrNativeEventWriter.hpp jfrEventWriterHost.inline.hpp jfrEventWriterHost.hpp jfrWriterHost.inline.hpp jfrTraceId.inline.hpp javaThread.inline.hpp oopHandle.inline.hpp access.inline.hpp barrierSet.inline.hpp objArrayOop.inline.hpp oop.inline.hpp 
  39x: shenandoahHeap.inline.hpp javaClasses.inline.hpp access.inline.hpp barrierSet.inline.hpp objArrayOop.inline.hpp oop.inline.hpp 
  32x: g1CollectedHeap.inline.hpp g1ConcurrentMark.inline.hpp g1ConcurrentMarkBitMap.inline.hpp markBitMap.inline.hpp oop.inline.hpp 
  30x: ciUtilities.inline.hpp interfaceSupport.inline.hpp javaThread.inline.hpp oopHandle.inline.hpp access.inline.hpp barrierSet.inline.hpp objArrayOop.inline.hpp oop.inline.hpp 
  ...

400304 ms: /jdk/src/hotspot/share/oops/oop.inline.hpp (included 1165 times, avg 343 ms), included via:
  80x: <direct include>
  70x: javaClasses.inline.hpp access.inline.hpp barrierSet.inline.hpp objArrayOop.inline.hpp 
  66x: oop.inline.hpp iterator.inline.hpp access.inline.hpp barrierSet.inline.hpp objArrayOop.inline.hpp 
  60x: javaClasses.inline.hpp access.inline.hpp barrierSet.inline.hpp objArrayOop.inline.hpp oop.inline.hpp iterator.inline.hpp instanceKlass.inline.hpp klass.inline.hpp classLoaderData.inline.hpp 
  40x: jfrEvents.hpp jfrEventClasses.hpp jfrEvent.hpp jfrNativeEventWriter.hpp jfrEventWriterHost.inline.hpp jfrEventWriterHost.hpp jfrWriterHost.inline.hpp jfrTraceId.inline.hpp javaThread.inline.hpp oopHandle.inline.hpp access.inline.hpp barrierSet.inline.hpp objArrayOop.inline.hpp 
  39x: shenandoahHeap.inline.hpp javaClasses.inline.hpp access.inline.hpp barrierSet.inline.hpp objArrayOop.inline.hpp 
  ...

[...]
```

This should give us a clear measure of what headers should go into `precompiled.hpp`, and should be more accurate than counting the number of inclusions. Now, improvements in build time are comparable with the initial approach I tried in this PR, but I think this approach will prove more accurate in the long term.

magicus commented:
That's interesting. Hopefully there is some consistency between compilers of how long time headers take to process. Ideally, we should get the same information from Microsoft CL, but I suspect that is not possible. (Since on Windows PCH makes the largest difference of any platform.)

shipilev commented on src/hotspot/share/precompiled/precompiled.hpp:
```diff
@@ -25,53 +25,27 @@
-#include "gc/shared/gcCause.hpp"
-#include "logging/log.hpp"
+#include "classfile/javaClasses.inline.hpp"
+#if INCLUDE_SHENANDOAHGC
```
So the common style is to move conditional includes at the end of the include block. I think the same rule applies here.

fandreuz replied:
Right, ba6f372d87765417f1bf05cd2b1a7e058c18350d

shipilev reviewed (commented):
Oh, `-ftime-trace` is cute.

Current PR still gives me considerable improvements on M1, I'd ballpark 15% faster Hotspot build.

```
CONF=macosx-aarch64-server-fastdebug make clean hotspot

# Before
1237.16s user 230.31s system 635% cpu 3:50.81 total
1239.11s user 233.62s system 643% cpu 3:48.88 total
1239.93s user 231.30s system 643% cpu 3:48.50 total

# After
1064.64s user 206.11s system 624% cpu 3:23.34 total
1067.86s user 208.56s system 621% cpu 3:25.39 total
1065.11s user 208.03s system 597% cpu 3:33.21 total
```

magicus commented:
There seems to be two similar arguments to `-ftime-trace` for cl.exe:  `/Bt+` and `/d2cgsummary`, according to my quick web search. It is not clear to me if they list include times. (But they might be useful in trying to figure out why files are slow to compile.)

magicus commented:
There is also https://github.com/microsoft/vcperf, which at a glance seems to give a lot of information. Not sure how hard it would be to integrate it into our build process, though...

fandreuz commented:
> The new PCH file is surprisingly short. But we should trust the performance measurements, and this seems to be much better than the existing file. It might be an indication that gcc/clang have improved in performance for non-PCH headers. If you are up to it, it would be interesting to have a comparison run on your test platforms for running without PCH. You don't have to retest everything you ran, but just some representative selection perhaps? It's not relevant to this PR, but more to the general discussion of how relevant it is to keep PCH at all.

@magicus it seems that PCH still makes quite a difference, I re-ran `server` builds:

| Branch  | Compiler | Build flavor   | Avg     | Stddev |
|---------|----------|-------|---------|--------|
| master  | gcc      | server no-pch| 3375.60 | 2.04   |
| master  | clang    | server  no-pch| 2554.08 | 6.25   |
| master   | gcc    | server   | 2962.90  | 3.04    |
| master   | clang  | server   | 2247.31  | 4.27    |
| 892ecb5a | gcc    | server   | 2570.87  | 0.96    |
| 892ecb5a | clang  | server   | 1824.21  | 2.31    |

The platform is the first in my [comment](https://github.com/openjdk/jdk/pull/26681#issuecomment-3219345092) above.

magicus commented:
No regression is acceptable, but I am still interested in seeing if adding back the `inline` files on Windows can give even better performance.

erikj79 commented:
> I added them back in [4a2ff56](https://github.com/openjdk/jdk/commit/4a2ff5649ba9dc6a024aeb7577d630a3fefcbe4f), there's only two of them since the others are now always included. @erikj79 any chance you could try once again and see if [4a2ff56](https://github.com/openjdk/jdk/commit/4a2ff5649ba9dc6a024aeb7577d630a3fefcbe4f) has any effect on build time?

I ran two more times:

This branch (@4a2ff5649ba9dc6a024aeb7577d630a3fefcbe4f): 4m37, 4m29

So basically unchanged.

fandreuz commented:
@magicus Thank you for all the feedback!

shipilev commented:
Ran the script locally, and noticed that on My mac this is was the difference:

```
-#include "oops/oopHandle.inline.hpp"
 #include "oops/oop.inline.hpp"
+#include "oops/oopHandle.inline.hpp"
```

Something is a bit off with sorting, I think.

Indeed, this fails:

```
$ diff --git a/test/hotspot/jtreg/sources/TestIncludesAreSorted.java b/test/hotspot/jtreg/sources/TestIncludesAreSorted.java
index 1a304a44944..287aad01b78 100644
--- a/test/hotspot/jtreg/sources/TestIncludesAreSorted.java
+++ b/test/hotspot/jtreg/sources/TestIncludesAreSorted.java
@@ -58,6 +58,7 @@ public class TestIncludesAreSorted {
                     "share/metaprogramming",
                     "share/oops",
                     "share/opto",
+                    "share/precompiled",
                     "share/services",
                     "share/utilities"
     };

$ CONF=macosx-aarch64-server-fastdebug make images test TEST=sources/
STDERR:
java.lang.RuntimeException: 1 files with unsorted headers found:
/Users/shipilev/Work/shipilev-jdk/src/hotspot/share/precompiled/precompiled.hpp
```

This is a minor thing, and we can technically deal with it once we sort the includes in `precompiled.hpp` in the constellation of issues that are dedicated to include sorting. But it would also be great to do this right from the get-go.

See if there is anything easy missing in the script?

fandreuz commented:
> See if there is anything easy missing in the script?

Thanks @shipilev, I addressed this in 6693cefe01e03cd05a11a93cf04d0d6af78f42bb. I use `SortIncludes.java` instead of `sort`, and I also added `precompiled` to `TestIncludesAreSorted.java`.

fandreuz commented on make/scripts/update_pch.sh:
```diff
@@ -0,0 +1,101 @@
+  >> "$PRECOMPILED_HPP.tmp"
+mv "$PRECOMPILED_HPP.tmp" "$PRECOMPILED_HPP"
+
+java /jdk/test/hotspot/jtreg/sources/SortIncludes.java --update /jdk/src/hotspot/share/precompiled
```
My bad, fixed in ead2c779b4481efb06b519924b5e56eeecad3674

That path is only valid in the Docker image where I tested the script, it shouldn't have been there.

fandreuz commented:
/integrate

shipilev commented:
I re-ran the script locally on my Mac, and I think it works without hitches now. There seems to be a tiny bit of variability for generated `precompiled.hpp`, since timing measurements are not exactly precise or consistent across build platforms. But IMO it is not worth diving deeper into.

The build times are still improving, as I tested earlier. 

So, great work all around here, folks. Kudos Francesco for trying multiple approaches and working out the great one! I think we are good to go. If we need any other changes, we can do them as follow-ups.

shipilev commented:
/sponsor

magicus commented:
Once again thank you @fandreuz for the effort you put into this!
